### PR TITLE
fix: use zones in the pre-existing frontend IP configurations for int…

### LINKED ISSUE #1047: Cloud provider should keep the AZ for ILB frontendIPConfiguration
#### What happened:

After #403, cloud provider would always set AZ when a region supports AZ. But because a region may switch from non-AZ to AZ, old  frontendIPConfiguration without AZ would get ResourceAvailabilityZonesCannotBeModified error when the cloud provider azure is upgraded to v0.7+ on the AZ supported regions.

#### What you expected to happen:

The old AZ values on ILB  frontendIPConfiguration should be kept when upgrading cloud provider Azure versions.

#### Environment:

- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by nilo19**
…ernal LBs

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Keep the AZ in the pre-existing frontend IP configurations of internal LBs instead of always calling the zone API.

#### Which issue(s) this PR fixes:

Fixes #1047

#### Does this PR introduce a user-facing change?

```release-note
fix: use zones in the pre-existing frontend IP configurations for internal LBs
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure_test.go [hunk 1]
```diff
@@ -1620,6 +1620,13 @@ func getInternalTestService(identifier string, requestedPorts ...int32) v1.Servi
 	svc.Annotations[consts.ServiceAnnotationLoadBalancerInternal] = consts.TrueAnnotationValue
 	return svc
 }
+func getInternalTestServiceWithAnnotation(identifier string, annotations map[string]string, requestedPorts ...int32) v1.Service {
```
Comment (MartinForReal) [inline pkg/provider/azure_test.go:1623]: How about implementing factory logic in getInternalTestServiceWithAnnotation and refactor getInternalTestService instead? Other functions may reuse logics in annotations without duplicating service generation logic.

File: pkg/provider/azure_loadbalancer.go [hunk 2]
```diff
@@ -1810,14 +1815,24 @@ func (az *Cloud) reconcileFrontendIPConfigs(clusterName string, service *v1.Serv
...
 
-			// only add zone information for new internal frontend IP configurations for standard load balancer not deployed to an edge zone.
-			location := az.Location
-			zones, err := az.getRegionZonesBackoff(location)
-			if err != nil {
-				return nil, false, err
-			}
-			if isInternal && az.useStandardLoadBalancer() && len(zones) > 0 && !az.HasExtendedLocation() {
-				newConfig.Zones = &zones
+			if !isFipChanged { // fetch zone information from API for new frontends
```
Comment (feiskyer) [inline pkg/provider/azure_loadbalancer.go:1818]: Could we wrap those new lines in a new function?

File: pkg/provider/azure_test.go [hunk 3]
```diff
@@ -1620,6 +1620,13 @@ func getInternalTestService(identifier string, requestedPorts ...int32) v1.Servi
 	svc.Annotations[consts.ServiceAnnotationLoadBalancerInternal] = consts.TrueAnnotationValue
 	return svc
 }
```
Comment (feiskyer) [inline pkg/provider/azure_test.go:1622]: gofmt the file? there should be an empty line between each functions

File: pkg/provider/azure_test.go (hunk 1 above)
Comment (feiskyer) [reply pkg/provider/azure_test.go:1623]: yes, that should help for new test suites

**Comment by feiskyer** [comment]
Could you fix the version in https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/azureclients/loadbalancerclient/interface.go#L29 together?

File: pkg/provider/azure_loadbalancer.go (hunk 2 above)
Comment (nilo19) [reply pkg/provider/azure_loadbalancer.go]: done

File: pkg/provider/azure_test.go (hunk 3 above)
Comment (nilo19) [reply pkg/provider/azure_test.go:1620]: done

**Comment by nilo19** [comment]
> Could you fix the version in https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/azureclients/loadbalancerclient/interface.go#L29 together?

fixed

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**